File: app/src/main/java/com/mmxb/mhelper/floatwindow/FloatWindow.java
package com.mmxb.mhelper.floatwindow;

import android.content.Context;
import android.graphics.Color;
import android.os.Build;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.view.WindowManager;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.mmxb.mhelper.MainWindow;
import com.mmxb.mhelper.R;

/**
 * Created by mmxb on 2019/2/20
 */
public class FloatWindow extends LinearLayout {

    // todo 弹性动画

    private WindowManager manager;
    private WindowManager.LayoutParams params;
    private int rawX;
    private int rawY;

    public FloatWindow(Context context) {
        this(context, null);
    }

    public FloatWindow(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public FloatWindow(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initView();
        initManager();
    }


    private void initView() {
        TextView textView = new TextView(getContext());
        textView.setText(R.string.helper);
        textView.setBackgroundColor(Color.GREEN);
        textView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                remove();
                new MainWindow(getContext()).show();
            }
        });
        addView(textView);
    }

    private void initManager() {
        manager = (WindowManager) getContext().getSystemService(Context.WINDOW_SERVICE);
        params = new WindowManager.LayoutParams();
        params.gravity = Gravity.END;
        params.width = WindowManager.LayoutParams.WRAP_CONTENT;
        params.height = WindowManager.LayoutParams.WRAP_CONTENT;
        //  params.flags 这两个属性很重要，避免了悬浮球全屏显示
        params.flags = WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL | WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            params.type = WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY;
        } else {
            params.type = WindowManager.LayoutParams.TYPE_PHONE;
        }
    }

    public void show() {
        manager.addView(this, params);
    }

    public void remove() {
        manager.removeView(this);
    }

    @Override
    public boolean dispatchTouchEvent(MotionEvent event) {
        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                rawX = (int) event.getRawX();
                rawY = (int) event.getRawY();
                break;
            case MotionEvent.ACTION_MOVE:
                params.x = params.x + (int) (event.getRawX() - rawX);
                params.y = params.y + (int) (event.getRawY() - rawY);
                rawX = (int) event.getRawX();
                rawY = (int) event.getRawY();
                manager.updateViewLayout(this, params);
                break;
            default:
                break;
        }
        return super.dispatchTouchEvent(event);
    }
}

File: app/src/main/java/com/mmxb/helper/floatwindow/FloatWindow.kt
package com.mmxb.helper.floatwindow

import android.content.Context
import android.graphics.PixelFormat
import android.os.Build
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.WindowManager
import android.widget.ImageView
import com.mmxb.helper.R
import com.mmxb.helper.main.ui.MainWindow
import com.mmxb.helper.util.ScreenUtil

/**
 * Created by mmxb on 2019/2/20
 */
class FloatWindow @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0)
    : ImageView(context, attrs, defStyleAttr) {

    private var manager = context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
    private var params = WindowManager.LayoutParams()
    private var rawX = 0
    private var rawY = 0
    private var startRawX = 0
    private var startRawY = 0

    init {
        initView()
        initManager()
    }


    private fun initView() {
        setOnClickListener {
            remove()
            MainWindow(context).show()
        }
        setImageResource(R.drawable.ic_float_window)
    }

    @Suppress("DEPRECATION")
    private fun initManager() {
        // 默认的gravity  Gravity.NO_GRAVITY ，坐标的原点在屏幕中间
        params.x = ScreenUtil.getSreenHeight(context) / 2
        params.y = ScreenUtil.getSreenHeight(context) / 5
        params.width = WindowManager.LayoutParams.WRAP_CONTENT
        params.height = WindowManager.LayoutParams.WRAP_CONTENT
        //  params.flags 这两个属性很重要，避免了悬浮球全屏显示
        params.flags = WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL or WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE
        //设置效果为背景透明.
        params.format = PixelFormat.RGBA_8888
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            params.type = WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY
        } else {
            params.type = WindowManager.LayoutParams.TYPE_PHONE
        }
    }

    fun show() {
        manager.addView(this, params)
    }

    fun remove() {
        manager.removeView(this)
    }

    override fun dispatchTouchEvent(event: MotionEvent): Boolean {
        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                rawX = event.rawX.toInt()
                rawY = event.rawY.toInt()
                startRawX = rawX
                startRawY = rawY
            }
            MotionEvent.ACTION_MOVE -> {
                val moveX = (event.rawX - rawX).toInt()
                val moveY = (event.rawY - rawY).toInt()
                params.x = params.x + moveX
                params.y = params.y + moveY
                rawX = event.rawX.toInt()
                rawY = event.rawY.toInt()
                manager.updateViewLayout(this, params)
            }
            MotionEvent.ACTION_UP -> {
                if (Math.abs(startRawX - rawX) > 8 || Math.abs(startRawY - rawY) > 8) {
                    scrollToSide()
                    // 消费掉事件，也就不会调用到view的onclick方法
                    return true
                } else {
                    return super.dispatchTouchEvent(event)
                }
            }
        }
        return super.dispatchTouchEvent(event)
    }

    private fun scrollToSide() {
        if (rawX > ScreenUtil.getSreenWidth(context) / 2) {
            params.x = ScreenUtil.getSreenWidth(context) / 2
        } else {
            params.x = -ScreenUtil.getSreenWidth(context) / 2
        }
        manager.updateViewLayout(this, params)
    }
}
